File: argon-plugins-js/shared/src/test/scala/dev/argon/plugins/js/ModuleResolution.scala
package dev.argon.plugins.js

import ModuleResolution.*
import zio.json.JsonDecoder
import zio.json.ast.{Json, JsonCursor}

import java.util.Locale

final class ModuleResolution(fileSystem: Map[String, String]) {

  val defaultConditions: Seq[String] = Seq("import")

  def resolveSpecifier(specifier: String, referencingModulePath: String): String =
    esmResolve(specifier, dirname(referencingModulePath))


  private def dirname(path: String): String =
    val index = path.lastIndexOf('/')
    if index > 0 then // Use > instead of >= to ensure that /file.txt results in /.
      path.substring(0, index).nn
    else
      "/"
  end dirname

  private def join(basePath: String, otherPath: String): String =
    if otherPath.startsWith("/") then
      otherPath
    else if otherPath.startsWith("./") then
      join(basePath, otherPath.substring(2).nn)
    else if otherPath.startsWith("../") then
      join(dirname(basePath), otherPath.substring(3).nn)
    else
      val index = otherPath.indexOf('/')
      if index >= 0 then
        join(basePath + "/" + otherPath.substring(0, index).nn, otherPath.substring(index + 1).nn)
      else
        s"$basePath/$otherPath"
    end if

  private def esmResolve(specifier: String, parentURL: String): String =
    val resolved =
      if specifier.startsWith("./") || specifier.startsWith("../") || specifier.startsWith("/") then
        join(parentURL, specifier)
      else if specifier.startsWith("#") then
        packageImportsResolve(specifier, parentURL, defaultConditions)
      else
        packageResolve(specifier, parentURL)

    if resolved.contains("%2F") || resolved.contains("%2f") || resolved.contains("%5C") || resolved.contains("%5c") then
      throw InvalidModuleSpecifierException(resolved)

    if fileSystem.keys.exists(_.startsWith(resolved + "/")) then
      throw UnsupportedDirectoryImportException(resolved)

    if !fileSystem.contains(resolved) then
      throw ModuleNotFoundException(resolved)

    resolved
  end esmResolve

  private def packageResolve(packageSpecifier: String, parentURL: String): String =
    val packageName =
      if packageSpecifier.isEmpty then
        throw InvalidModuleSpecifierException(packageSpecifier)
      else if !packageSpecifier.startsWith("@") then
        val index = packageSpecifier.indexOf('/')
        if index >= 0 then
          packageSpecifier.substring(0, index).nn
        else
          packageSpecifier
      else
        val index = packageSpecifier.indexOf('/')
        if index >= 0 then
          val index2 = packageSpecifier.indexOf('/', index + 1)
          if index2 >= 0 then
            packageSpecifier.substring(0, index2).nn
          else
            packageSpecifier
        else
          throw InvalidModuleSpecifierException(packageSpecifier)
        end if
      end if

    if packageName.startsWith(".") || packageName.contains("\\") || packageName.contains("%") then
      throw InvalidModuleSpecifierException(packageSpecifier)



    val packageSubpath = "." + packageSpecifier.substring(packageName.length)

    if packageSubpath.endsWith("/") then
      throw InvalidModuleSpecifierException(packageSpecifier)


    packageSelfResolve(packageName, packageSubpath, parentURL).getOrElse {
      def loop(parentURL: String): String =
        if parentURL == "/" then
          throw ModuleNotFoundException(packageSpecifier)
        else
          val packageURL = join(parentURL, "node_modules/" + packageName)
          readPackageJson(packageURL)
            .map { pjson =>
              def packageExportPath: Option[String] =
                pjson.get(JsonCursor.isObject.field("exports"))
                  .toOption
                  .filterNot { _.isInstanceOf[Json.Null.type] }
                  .map { pjsonExports => packageExportsResolve(packageURL, packageSubpath, pjsonExports, defaultConditions) }

              def packageMain: Option[String] =
                if packageSubpath == "." then
                  pjson.get(JsonCursor.isObject.field("main").isString)
                    .map { _.value }
                    .toOption
                else
                  None

              packageExportPath.orElse(packageMain).getOrElse(join(packageURL, packageSubpath))
            }
            .getOrElse(loop(dirname(parentURL)))
        end if
      end loop

      loop(parentURL.stripSuffix("/"))
    }
  end packageResolve

  private def packageSelfResolve(packageName: String, packageSubpath: String, parentURL: String): Option[String] =
    for
      packageURL <- lookupPackageScope(parentURL)
      pjson <- readPackageJson(packageURL)
      pjsonName <- pjson.get(JsonCursor.isObject.field("name").isString).toOption if pjsonName.value == packageName
      pjsonExports <- pjson.get(JsonCursor.isObject.field("exports")).toOption if !pjsonExports.isInstanceOf[Json.Null.type]
    yield packageExportsResolve(packageURL, packageSubpath, pjsonExports, defaultConditions)

  def packageExportsResolve(packageURL: String, subpath: String, exports: Json, conditions: Seq[String]): String =
    val exportsWithDotKey = exports match {
      case obj @ Json.Obj(fields) if fields.exists { _._1.startsWith(".") } => Some(obj)
      case _ => None
    }
    val exportsWithNonDotKey = exports match {
      case obj @ Json.Obj(fields) if fields.exists { !_._1.startsWith(".") } => Some(obj)
      case _ => None
    }

    if exportsWithDotKey.isDefined && exportsWithNonDotKey.isDefined then
      throw InvalidPackageConfigurationException(packageURL)

    if subpath == "." then
      exports.get(JsonCursor.isString).toOption
        .orElse(exports.get(JsonCursor.isArray).toOption)
        .orElse(exports.get(JsonCursor.isObject).toOption.filter { _ => exportsWithDotKey.isEmpty })
        .orElse(exports.get(JsonCursor.isObject.field(".")).toOption)
        .flatMap { mainExport => packageTargetResolve(packageURL, mainExport, "", false, false, conditions) }
        .getOrElse { throw PackagePathNotExportedException(packageURL, ".") }
    else
      exportsWithDotKey
        .flatMap { exports =>
          packageImportsExportsResolve(subpath, exports, packageURL, false, conditions)
        }
        .getOrElse {
          throw PackagePathNotExportedException(packageURL, subpath)
        }
    end if
  end packageExportsResolve

  private def packageImportsResolve(specifier: String, parentURL: String, conditions: Seq[String]): String =
    if !specifier.startsWith("#") || specifier == "#" || specifier.startsWith("#/") then
      throw InvalidModuleSpecifierException(specifier)

    (
      for
        packageURL <- lookupPackageScope(parentURL)
        pjson <- readPackageJson(packageURL)
        pjsonImports <- pjson.get(JsonCursor.isObject.field("imports").isObject).toOption
        resolved <- packageImportsExportsResolve(specifier, pjsonImports, packageURL, true, conditions)
      yield resolved
    ).getOrElse { throw PackageImportNotDefinedException(specifier) }

  private def packageImportsExportsResolve(matchKey: String, matchObj: Json.Obj, packageURL: String, isImports: Boolean, conditions: Seq[String]): Option[String] =
    (
      if matchKey.contains("*") then
        None
      else
        matchObj.get(JsonCursor.isObject.field(matchKey))
          .map { target => packageTargetResolve(packageURL, target, "", false, isImports, conditions) }
          .toOption
    ).getOrElse {
      def keyHasSingleStar(key: String): Boolean =
        val first = key.indexOf('*')
        first >= 0 && key.indexOf('*', first + 1) < 0
      end keyHasSingleStar

      def patternKeyLessThan(keyA: String, keyB: String): Boolean =
        val baseLengthA =
          val index = keyA.indexOf('*')
          if index >= 0 then index + 1
          else keyA.length
        end baseLengthA

        val baseLengthB =
          val index = keyB.indexOf('*')
          if index >= 0 then index + 1
          else keyB.length
        end baseLengthB

        if baseLengthA > baseLengthB then
          true
        else if baseLengthB > baseLengthA then
          false
        else if baseLengthA == keyA.length then
          false
        else if baseLengthB == keyB.length then
          true
        else if keyA.length > keyB.length then
          true
        else
          false
      end patternKeyLessThan

      matchObj.fields
        .filter { case (k, _) => keyHasSingleStar(k) }
        .sortWith { case ((k1, _), (k2, _)) => patternKeyLessThan(k1, k2) }
        .collectFirst(({ (expansionKey, target) =>
          val starIndex = expansionKey.indexOf('*')
          val patternBase = expansionKey.substring(0, starIndex).nn
          if matchKey.startsWith(patternBase) && matchKey != patternBase then
            val patternTrailer = expansionKey.substring(starIndex + 1)
            if matchKey.endsWith(patternTrailer) && matchKey.length >= expansionKey.length then
              val subpath = matchKey.substring(starIndex, matchKey.length - expansionKey.length).nn
              packageTargetResolve(packageURL, target, subpath, true, isImports, conditions)
            else
              None
            end if
          else
            None
          end if
        } : ((String, Json)) => Option[String]).unlift)
    }

  def packageTargetResolve(packageURL: String, target0: Json, subpath: String, pattern: Boolean, internal: Boolean, conditions: Seq[String]): Option[String] =
    target0 match {
      case Json.Str(target) =>
        if !pattern && subpath.nonEmpty && !target.endsWith("/") then
          throw InvalidModuleSpecifierException(subpath)

        val result =
          if target.startsWith("./") then
            def isInvalidSeg(seg: String): Boolean =
              seg == "." || seg == ".." || seg.toLowerCase(Locale.US).nn == "node_modules"

            if target.split(Array('/', '\\')).drop(1).exists(isInvalidSeg) then
              throw InvalidPackageTargetException(target0)

            val resolvedTarget = join(packageURL, target)

            if subpath.split(Array('/', '\\')).exists(isInvalidSeg) then
              throw InvalidModuleSpecifierException(subpath)

            if pattern then
              resolvedTarget.replace("*", subpath).nn
            else
              join(subpath, resolvedTarget)

          else
            if internal && !target.startsWith("../") && !target.startsWith("/") then
              if pattern then
                packageResolve(target.replace("*", subpath).nn, packageURL + "/")
              else
                packageResolve(target + subpath, packageURL + "/")
            else
              throw InvalidPackageTargetException(target0)
          end if

        Some(result)

      case Json.Obj(fields) =>
        fields
          .iterator
          .filter { (key, _) => key == "default" || conditions.contains(key) }
          .collectFirst((((key, value) =>
            packageTargetResolve(packageURL, value, subpath, pattern, internal, conditions)
          ) : ((String, Json)) => Option[String]).unlift)

      case Json.Arr(elements) =>
        elements.collectFirst(((targetValue =>
          packageTargetResolve(packageURL, targetValue, subpath, pattern, internal, conditions)
        ) : Json => Option[String]).unlift)

      case _: Json.Null.type =>
        None

      case _ =>
        throw InvalidPackageTargetException(target0)
    }

  private def lookupPackageScope(url: String): Option[String] =
    if url == "/" then
      None
    else
      val scopeURL = dirname(url)
      if scopeURL.endsWith("/node_modules") then
        None
      else if fileSystem.contains(join(scopeURL, "package.json")) then
        Some(scopeURL)
      else
        None
    end if

  private def readPackageJson(packageURL: String): Option[Json] =
    fileSystem.get(join(packageURL, "package.json"))
      .map { jsonText =>
        summon[JsonDecoder[Json]].decodeJson(jsonText)
          .getOrElse { throw InvalidPackageConfigurationException(packageURL) }
      }




}

object ModuleResolution {
  final class InvalidModuleSpecifierException(specifier: String) extends Exception(s"Invalid specifier: $specifier")
  final class ModuleNotFoundException(path: String) extends Exception(s"Could not find module: $path")
  final class UnsupportedDirectoryImportException(path: String) extends Exception
  final class InvalidPackageConfigurationException(packageURL: String) extends Exception
  final class PackagePathNotExportedException(packageURL: String, subpath: String) extends Exception(s"Module $packageURL does not export path $subpath")
  final class PackageImportNotDefinedException(specifier: String) extends Exception
  final class InvalidPackageTargetException(target: Json) extends Exception
}

File: backend/backends/js/src/executor/moduleResolution.ts
import type * as estree from "estree";
import type { PackageJson, ReadonlyDeep } from "type-fest";

export interface ESMResolution {
    resolved: URL;
    format: string | undefined;
}

export class ModuleResolutionError extends Error {}
export class NonFileUrlError extends ModuleResolutionError {}
export class InvalidModuleSpecifierError extends ModuleResolutionError {}
export class UnsupportedDirectoryImport extends ModuleResolutionError {}
export class ModuleNotFoundError extends ModuleResolutionError {}
export class PackagePathNotExportedError extends ModuleResolutionError {}
export class PackageImportNotDefinedError extends ModuleResolutionError {}
export class InvalidPackageTargetError extends ModuleResolutionError {}
export class InvalidPackageConfigurationError extends ModuleResolutionError {}


export class ModuleResolution {

    defaultConditions: string[] = ["import"];
    
    private packageJsonFiles = new Map<string, ReadonlyDeep<PackageJson>>();
    private sourceFiles = new Map<string, ReadonlyDeep<estree.Program>>();
    private directories = new Set<string>([ "/" ]);


    addSourceFile(path: string, content: ReadonlyDeep<estree.Program>): void {
        if(!path.startsWith("/")) {
            path = "/" + path;
        }

        this.sourceFiles.set(path, content);
        this.addDirectory(this.getPathParent(path));
    }

    addPackageJsonFile(path: string, content: ReadonlyDeep<PackageJson>): void {
        if(!path.startsWith("/")) {
            path = "/" + path;
        }

        this.packageJsonFiles.set(path, content);
        this.addDirectory(this.getPathParent(path));
    }

    addDirectory(path: string): void {
        while(path.endsWith("/")) {
            path = path.substring(0, path.length - 1);
        }

        if(!path.startsWith("/")) {
            path = "/" + path;
        }

        while(path.length > 1) {
            this.directories.add(path);
            path = this.getPathParent(path);
        }
    }


    esmResolve(specifier: string, parentURL: URL): ESMResolution {
        const resolved = (() => {
            const specUrl = URL.parse(specifier);
            if(specUrl != null) {
                return specUrl;
            }
    
            if(specifier.startsWith("/") || specifier.startsWith("./") || specifier.startsWith("../")) {
                return new URL(specifier, parentURL);
            }
            else if(specifier.startsWith("#")) {
                return this.packageImportsResolve(specifier, parentURL, this.defaultConditions);
            }
            else {
                return this.packageResolve(specifier, parentURL);
            }
        })();
    
        if(resolved.protocol !== "file:") {
            throw new NonFileUrlError(`Protocols other than file are not supported: ${resolved.protocol}`);
        }

        const resolvedStr = resolved.toString();
        if(resolvedStr.indexOf("%2F") >= 0 || resolvedStr.indexOf("%2f") >= 0 || resolvedStr.indexOf("%5C") >= 0 || resolvedStr.indexOf("%5c") >= 0) {
            throw new InvalidModuleSpecifierError(`Specifier contains invalid characters ${resolvedStr}`);
        }

        if(this.isDirectory(resolved.pathname)) {
            throw new UnsupportedDirectoryImport(`Directory cannot be imported: ${resolved}`);
        }

        if(!this.sourceFiles.has(resolved.pathname)) {
            throw new ModuleNotFoundError(`Module not found at ${resolved}`);
        }

        const format = this.esmFileFormat(resolved);

        return { resolved, format };
    }

    packageResolve(packageSpecifier: string, parentURL: URL): URL {
        if(packageSpecifier === "") {
            throw new InvalidModuleSpecifierError("Package specifier was empty");
        }

        // Not checking for node builtin modules because we won't have any.

        let packageName: string;
        if(packageSpecifier.startsWith("@")) {
            const firstSlash = packageSpecifier.indexOf("/");
            if(firstSlash < 0) {
                throw new InvalidModuleSpecifierError();
            }

            const secondSlash = packageSpecifier.indexOf("/", firstSlash + 1);
            if(secondSlash < 0) {
                packageName = packageSpecifier;
            }
            else {
                packageName = packageSpecifier.substring(0, secondSlash);
            }
        }
        else {
            const slash = packageSpecifier.indexOf("/");
            if(slash < 0) {
                packageName = packageSpecifier;
            }
            else {
                packageName = packageSpecifier.substring(0, slash);
            }
        }

        if(packageName.startsWith(".") || packageName.indexOf("\\") >= 0 || packageName.indexOf("%") >= 0) {
            throw new InvalidModuleSpecifierError("Package name contains invalid characters: " + packageName);
        }

        const packageSubpath = "." + packageSpecifier.substring(packageName.length);
        if(packageSubpath.endsWith("/")) {
            throw new InvalidModuleSpecifierError();
        }

        const selfUrl = this.packageSelfResolve(packageName, packageSubpath, parentURL);
        if(selfUrl !== undefined) {
            return selfUrl;
        }

        while(!this.isRootUrl(parentURL)) {
            const packageURL = new URL("node_modules/" + packageName + "/", parentURL);
            parentURL = new URL("..", parentURL);

            if(!this.isDirectory(packageURL.pathname)) {
                continue;
            }

            const pjson = this.readPackageJson(packageURL);
            if(pjson !== null) {
                if(pjson.exports !== undefined && pjson.exports !== null) {
                    return this.packageExportsResolve(packageURL, packageSubpath, pjson.exports, this.defaultConditions);
                }
            }

            if(packageSubpath === ".") {
                if(typeof pjson?.main === "string") {
                    return new URL(pjson.main, packageURL);
                }
            }

            return new URL(packageSubpath, packageURL);
        }

        throw new ModuleNotFoundError("Could not find module: " + packageSpecifier);
    }

    packageSelfResolve(packageName: string, packageSubpath: string, parentURL: URL): URL | undefined {
        const packageURL = this.lookupPackageScope(parentURL);
        if(packageURL === null) {
            return undefined;
        }

        const pjson = this.readPackageJson(packageURL);
        if(pjson === null || pjson.exports === null || pjson.exports === undefined) {
            return undefined;
        }

        if(pjson.name === packageName) {
            return this.packageExportsResolve(packageURL, packageSubpath, pjson.exports, this.defaultConditions);
        }

        return undefined;
    }

    packageExportsResolve(packageURL: URL, subpath: string, exports: {}, conditions: readonly string[]): URL {
        let hasDot = false;
        let hasNonDot = false;
        if(typeof exports === "object" && !(exports instanceof Array)) {
            const keys = Object.keys(exports);
            for(const key of keys) {
                if(key.startsWith(".")) {
                    hasDot = true;
                }
                else {
                    hasNonDot = true;
                }
            }

            if(hasDot && hasNonDot) {
                throw new InvalidModuleSpecifierError("Exports object contains entries both starting with and without a dot.");
            }
        }

        if(subpath === ".") {
            let mainExport: unknown = undefined;
            if(typeof exports === "string" || exports instanceof Array || (typeof exports === "object" && !hasDot)) {
                mainExport = exports;
            }
            else if(typeof exports === "object" && exports !== null && "." in exports) {
                mainExport = exports["."];
            }

            if(mainExport !== undefined) {
                const resolved = this.packageTargetResolve(packageURL, mainExport, null, false, conditions);
                if(resolved !== null && resolved !== undefined) {
                    return resolved;
                }
            }
        }
        else if(hasDot) {
            const resolved = this.packageImportsExportsResolve(subpath, exports, packageURL, false, conditions);
            if(resolved !== null && resolved !== undefined) {
                return resolved;
            }
        }

        throw new PackagePathNotExportedError("Subpath was not exported: " + subpath);
    }

    packageImportsResolve(specifier: string, parentURL: URL, conditions: readonly string[]): URL {
        if(specifier === "#" || specifier.startsWith("#/")) {
            throw new InvalidModuleSpecifierError("Invalid import specifier: " + specifier);
        }

        const packageURL = this.lookupPackageScope(parentURL);
        if(packageURL != null) {
            const pjson = this.readPackageJson(packageURL);

            if(typeof pjson === "object" && pjson != null && "imports" in pjson && typeof pjson.imports === "object" && pjson.imports !== null) {
                const resolved = this.packageImportsExportsResolve(specifier, pjson.imports, packageURL, true, conditions);
                if(resolved !== null && resolved !== undefined) {
                    return resolved;
                }
            }
        }

        throw new PackageImportNotDefinedError("Import specifier was not defined: " + specifier);
    }

    packageImportsExportsResolve(matchKey: string, matchObj: object, packageURL: URL, isImports: boolean, conditions: readonly string[]): URL | undefined | null {
        if(typeof matchObj !== "object" || matchObj === null) {
            return null;
        }

        if(matchKey in matchObj && matchKey.indexOf("*") < 0) {
            const target: unknown = (matchObj as any)[matchKey];

            if(target !== undefined) {
                return this.packageTargetResolve(packageURL, target, null, isImports, conditions);
            }
        }

        const expansionKeys = Object.keys(matchObj)
            .filter(k => {
                const firstIndex = k.indexOf("*");
                if(firstIndex < 0) {
                    return false;
                }

                return k.indexOf("*", firstIndex + 1) < 0;
            })
            .sort((keyA, keyB) => {
                const baseLengthA = keyA.indexOf("*");
                const baseLengthB = keyB.indexOf("*");
                if(baseLengthA > baseLengthB) return -1;
                if(baseLengthB > baseLengthA) return 1;
                if(keyA.length > keyB.length) return -1;
                if(keyB.length > keyA.length) return -1;
                return 0;
            });

        for(const expansionKey of expansionKeys) {
            const starIndex = expansionKey.indexOf("*");
            const patternBase = expansionKey.substring(0, starIndex);
            if(matchKey.startsWith(patternBase) && matchKey !== patternBase) {
                const patternTrailer = expansionKey.substring(starIndex + 1);
                if(patternTrailer.length === 0 || (matchKey.endsWith(patternTrailer) && matchKey.length >= expansionKey.length)) {
                    const target: unknown = (matchObj as any)[matchKey];
                    const patternMatch = matchKey.substring(patternBase.length, matchKey.length - patternTrailer.length);
                    return this.packageTargetResolve(packageURL, target, patternMatch, isImports, conditions);
                }
            }
        }

        return null;
    }

    packageTargetResolve(packageURL: URL, target: unknown, patternMatch: string | null, isImports: boolean, conditions: readonly string[]): URL | undefined | null {
        if(typeof target === "string") {
            if(!target.startsWith("./")) {
                if(!isImports || target.startsWith("../") || target.startsWith("/") || URL.canParse(target)) {
                    throw new InvalidPackageTargetError("Invalid package target: " + target);
                }

                if(patternMatch !== null) {
                    return this.packageResolve(target.replaceAll("*", patternMatch), new URL(packageURL.toString() + "/"));
                }

                return this.packageResolve(target, packageURL);
            }

            if(this.hasBadSegment(target.substring(2))) {
                throw new InvalidModuleSpecifierError("Invalid module specifier: " + target);
            }

            const resolvedTarget = new URL(target, packageURL);
            if(patternMatch === null) {
                return resolvedTarget;
            }

            if(this.hasBadSegment(patternMatch)) {
                throw new InvalidModuleSpecifierError("Invalid module specifier: " + patternMatch);
            }

            return new URL(resolvedTarget.toString().replaceAll("*", patternMatch));
        }
        else if(target === null) {
            return null;
        }
        else if(target instanceof Array) {
            let res: InvalidPackageTargetError | undefined | null = null;
            
            for(const targetValue of target) {
                try {
                    const resolved = this.packageTargetResolve(packageURL, targetValue, patternMatch, isImports, conditions);
                    if(resolved === undefined) {
                        res = resolved;
                        continue;
                    }
                    return resolved;
                }
                catch(e) {
                    if(e instanceof InvalidPackageTargetError) {
                        res = e;
                        continue;
                    }
                    throw e;
                }
            }

            if(res === undefined || res === null) {
                return res;
            }

            throw res;
        }
        else if(typeof target === "object") {
            if(Object.keys(target).some(k => this.isNumericKey(k))) {
                throw new InvalidPackageConfigurationError();
            }

            for(const [p, targetValue] of Object.entries(target)) {
                if(p === "default" || conditions.indexOf(p) >= 0) {
                    const resolved = this.packageTargetResolve(packageURL, targetValue, patternMatch, isImports, conditions);
                    if(resolved === undefined) {
                        continue;
                    }

                    return resolved;
                }
            }

            return undefined;
        }
        else {
            throw new InvalidPackageTargetError("Invalid package target: " + JSON.stringify(target));
        }
    }

    esmFileFormat(url: URL): string | undefined {
        const s = url.toString();

        if(s.endsWith(".mjs")) return "module";
        if(s.endsWith(".cjs")) return "commonjs";
        if(s.endsWith(".json")) return "json";

        const packageUrl = this.lookupPackageScope(url);
        if(packageUrl === null) {
            return undefined;
        }

        const pjson = packageUrl == null ? null : this.readPackageJson(packageUrl);

        let packageType: string | null = null;
        if(pjson?.type === "module" || pjson?.type === "commonjs") {
            packageType = pjson.type;
        }
        
        if(s.endsWith(".js")) {
            if(packageType !== null) {
                return packageType;
            }

            if(this.detectModuleSyntax(url)) {
                return "module";
            }
            else {
                return "commonjs";
            }
        }

        if(s.match(/\.[^\.\\/]+$/)) {
            if(this.detectModuleSyntax(url)) {
                return "module";
            }
            else {
                return "commonjs";
            }
        }

        return undefined;        
    }

    lookupPackageScope(url: URL): URL | null {
        let scopeURL = url;

        while(!this.isRootUrl(scopeURL)) {
            if(scopeURL.pathname.endsWith("/node_modules/") || scopeURL.pathname.endsWith("/node_modules")) {
                return null;
            }

            const pjsonURL = new URL("package.json", scopeURL);
            if(this.packageJsonFiles.has(pjsonURL.pathname)) {
                return scopeURL;
            }

            scopeURL = new URL("..", scopeURL);
        }

        return null;
    }

    readPackageJson(packageURL: URL): ReadonlyDeep<PackageJson> | null {
        const pjsonURL = new URL("package.json", packageURL);
        if(!this.packageJsonFiles.has(pjsonURL.pathname)) {
            return null;
        }

        return this.packageJsonFiles.get(pjsonURL.pathname)!;
    }

    private hasBadSegment(s: string): boolean {
        const segments = s.split(/\/|\\/);
        return segments.some(seg => {
            seg = decodeURIComponent(seg).toLowerCase();
            return seg === "" || seg === "." || seg === ".." || seg === "node_modules";
        });
    }



    // stubbing for now
    detectModuleSyntax(_url: URL): boolean {
        return true;
    }


    private isDirectory(path: string): boolean {
        if(path !== "/" && path.endsWith("/")) {
            path = path.substring(0, path.length - 1);
        }

        return this.directories.has(path);
    }

    private getPathParent(path: string): string {
        const index = path.lastIndexOf("/");
        if(index > 0) {
            return path.substring(0, index);
        }
        else {
            return "/";
        }
    }

    private isRootUrl(url: URL): boolean {
        return url.toString() == new URL("..", url).toString();
    }

    private isNumericKey(k: string): boolean {
        let n: bigint;
        try {
            n = BigInt(k);
        }
        catch(_) {
            return false;
        }

        if(k !== n.toString()) {
            return false;
        }

        return n >= BigInt(Number.MIN_SAFE_INTEGER) && n <= BigInt(Number.MAX_SAFE_INTEGER);
    }
}


